### ISSUE TITLE
feat: Updates to  ludwig configuration for improved readability.

### ISSUE THREAD

**Issue by brightsparc** [ludwig-ai/ludwig#1738]
**Is your feature request related to a problem? Please describe.**
No, this feature request relates to improving the readability of the configuration.

**Describe the use case**
This feature request is to propose some changes the the ludwig yaml configuration which make explicit the encoder and decoder types and properties, and unnest the preprocessing configuration to be included at the same level of the input/output feature.

**Describe the solution you'd like**
I propose the following changes for properties in ludwig configuration:
1. New top level `version` attribute in config file.
2. Rename the `numerical` type to be `numeric`.
3. Move `encoder` and `decoder` properties to be nested under input and output features, with a `type` property.
4. Rename the `preprocessing` section to be `processor` given it is not specific to pre or post processing.
5. Rename the `training` section to be `trainer` to be consistent with the `er` suffiix.

The following sample configuration provides an example of what the changes would look like:

```
version: 0.5
processor:
    numeric:
        missing_value_strategy: fill_with_mean
input_features:
    -
        name: numerical_column_name
        type: numeric
        encoder:
             type: dense
             num_layers: 1
        processor:
             missing_value_strategy: fill_value
             fill_value: -1
    -
        name: binary_column_name
        type: binary
        encoder: 
            type: passthrough
output_features:
    -
        name: class_name
        type: category
        decoder:
             loss:
                  weight: 1.5
combiner:
   type: concat
trainer:
   epochs: 200
```

Given this change will touch a large section of the code based, I am proposing we complete this before other breaking changes in 0.5.  To provide a smooth upgrade path, I propose we include a `version` attribute in the file, and a utility to support upgrade files that are not at the latest version 0.5

**Describe alternatives you've considered**
The alternative is not change the configuration file.

**Additional context**
For comparison, this would the original version of the above configuration file.

```
preprocessing:
    numerical:
        missing_value_strategy: fill_with_mean
input_features:
    -
        name: numerical_column_name
        type: numerical
        encoder: dense
        num_layers: 1
        preprocessing:
             missing_value_strategy: fill_value
             fill_value: -1
    -
        name: binary_column_name
        type: binary
        encoder: passthrough
output_features:
    -
        name: class_name
        type: category
        loss:
              weight: 1.5
combiner:
   type: concat
training:
   epochs: 200
```

**Comment by justinxzhao** [comment]
Putting in a quick vote for going back to `preprocessing`, the process, over `processor`:

- Parameters in this block, like `fill_with_mean`, are predominantly associated with the data "preprocessing" aspect of the ML pipeline.
- A few parameters like `str2idx` or `start_symbol` are indeed referenced in post-processing, but only because they were used in pre-processing, rather than being specifically necessary for post-processing.
- At standup yesterday, @ShreyaR also brought up a potential use case with ranking after scoring that we may want to enable configuring post-processing-specific parameters in the future.
- "Data preprocessing" is widely used terminology. It feels odd to depart from that convention.
- Data preprocessing can be sharded across many machines, and `preprocessing` makes it clearer that we're configuring the process, rather than a specific object.

The last bullet is also why I have a preference (though slighter) for `training` instead of `trainer`, as in we're configuring the training process, rather than a specific Trainer, which is potentially instantiated on and coordinated across several machines.

```
training:
    checkpoint_steps: 2000
    epochs: 200
    tree:
        criterion: gini
        splitter: best
        max_depth: 10
        min_samples_split: 2
        min_samples_leaf: 1
        min_weight_fraction_leaf: 0.0
        ...
    ecd:
        learning_rate: 0.001
        batch_size: 8196
        ...
```

**Comment by brightsparc** [comment]
Proposed changes from @w4nderlust 

* numerical -> number. I prefer it over numeric because it’s a noun. Also considered numeric but in sql (at least some dialects) that’ an exact value, so it could be confusing. Also considered real, but that also may imply a certain level of approximate precision. I think number is a more neutral and general choice.
* training -> trainer. The main reason for this is that I’m imagining other trainers with a type in the future (like ssl_trainer or unsupervised-trainer) which may have a different set of parameters. I see the point of considering it a shorthand for training process, but that wouldn’t work as well in a future with multiple trainers

These changes are closed by PR #1743
